opm/simulators/wells/WellGroupHelpers.hpp

```diff
@@ -0,0 +1,211 @@
+                continue;
+
+            double factor = wellEcl.getEfficiencyFactor();
+            const auto wellrate_index = well_index * wellState.numPhases();
+            if (injector)
+                rate += factor * wellState.wellRates()[ wellrate_index + phasePos];
+            else
+                rate -= factor * wellState.wellRates()[ wellrate_index + phasePos];
+        }
+    }
+
+    inline void sumWellResRates(const Group2& group, const Schedule& schedule, const WellStateFullyImplicitBlackoil& wellState, const int reportStepIdx, const int phasePos, const bool injector, double& rate) {
+
+        for (const std::string& groupName : group.groups()) {
+            const Group2& groupTmp = schedule.getGroup2(groupName, reportStepIdx);
+            sumWellResRates(groupTmp, schedule, wellState, reportStepIdx, phasePos, injector, rate);
+            rate *= groupTmp.getGroupEfficiencyFactor();
+        }
+        const auto& end = wellState.wellMap().end();
+        for (const std::string& wellName : group.wells()) {
+            const auto& it = wellState.wellMap().find( wellName );
+            if (it == end)  // the well is not found
+                continue;
+
+            int well_index = it->second[0];
+
+            const auto& wellEcl = schedule.getWell2(wellName, reportStepIdx);
+            //only count producers or injectors
+            if ( (wellEcl.isProducer() && injector) ||  (wellEcl.isInjector() && !injector))
+                continue;
+
+            double factor = wellEcl.getEfficiencyFactor();
+
+            const auto wellrate_index = well_index * wellState.numPhases();
+            if (injector)
+                rate += factor * wellState.wellReservoirRates()[ wellrate_index + phasePos];
+            else
+                rate -= factor * wellState.wellReservoirRates()[ wellrate_index + phasePos];
+        }
+    }
+
+    inline void setGroupControl(const Group2& group, const Schedule& schedule, const int reportStepIdx, const bool injector, WellStateFullyImplicitBlackoil& wellState) {
+
+        for (const std::string& groupName : group.groups()) {
+            const Group2& groupTmp = schedule.getGroup2(groupName, reportStepIdx);
+            setGroupControl(groupTmp, schedule, reportStepIdx, injector, wellState);
+            if (injector)
+                wellState.setCurrentInjectionGroupControl(groupName, "FLD");
+            else
+                wellState.setCurrentProductionGroupControl(groupName, "FLD");
+        }
+
+        const auto& end = wellState.wellMap().end();
+        for (const std::string& wellName : group.wells()) {
+            const auto& it = wellState.wellMap().find( wellName );
+            if (it == end)  // the well is not found
+                continue;
+
+            int well_index = it->second[0];
+            const auto& wellEcl = schedule.getWell2(wellName, reportStepIdx);
+            if (!wellEcl.isAvailableForGroupControl())
```

Comment:
It seems insufficient to check only the static information here, if the well is at its own limit it should not be available for group control?

Reply:
This is correct as is. See comment above.

Reply:
I agree, it should switch back if necessary.